Green audio player

mzgoddard opened the pull request:
- [x] Depends on #88.

### Proposed changes

Use GreenPlay in AudioPlayer.

### Reason for changes

- Use the new SoundPlayer API in AudioPlayer.
- Support stopping already playing sounds without clipping through SoundPlayer's behaviour.

mzgoddard committed:
feat: use GreenPlayer in AudioPlayer

gnarf commented:
What's the commit range on this that isn't part of #88 ? can you link to the pr diff with just that commits?

mzgoddard commented:
Cleaned up the conflict. Should be ready for review now.

mzgoddard committed:
add GreenPlayer.isStarting promise

Self-resolving promise removes itself once it resolves. While it is on
a player the player may consider itself to not yet have the time to
actually start sound playback. While playback has not started, the
player can shortcut out of starting playback.

gnarf commented on src/GreenPlayer.js:
```diff
@@ -198,6 +205,13 @@ class SoundPlayer extends EventEmitter {
 
         this.isPlaying = true;
 
+        const isStarting = this.isStarting = Promise.resolve()
```
hrm... part of me thinks it might be better to just compare the lastStartTime to audioContext.currentTime -- this kind of depends on the microtasks running in the right order and might still emit a bunch of start events at the same audio context time?

mzgoddard replied:
This needs to be aware of the javascript context's state, has the rest of the main thread performed work. This is to keep users from playing a possibly infinite number of copies of a sound.

This could test both things but I don't think it can work only by testing time. The time it cares about is at the millisecond scale anyways.

(We may want to make a separate microsecond check to determine if we should take and stop or take and stopImmediately.)

mzgoddard committed:
use AudioContext.currentTime + DECAY_TIME to debounce sound start

Debounce when the sound starts to keep from playing many copies of the
same sound layered on itself. Use AudioEngine's currentTime in seconds
plus some value in milliseconds to make sure the sound has a chance to
start before starting a second playback.

ericrosenbaum reviewed (approved):
🎉
